gui/console.cpp

```diff
@@ -620,13 +620,27 @@ void ConsoleDialog::saveHistory() {
 		return;
 	}
 
-	for (int i = 0; i < _historySize; ++i) {
-		saveFile->writeString(_history[i]);
-		saveFile->writeByte('\n');
+	// Saving the history entries in the proper order;
+	// The most recent entry should be the last to be saved.
+	// NOTE When the _history table is full, we need to start saving
+	//      from one slot after (in a circular manner) the _historyIndex slot.
+	//      In this case the _historyIndex slot contains the temporary stored user input,
+	//      which we do not want to persist.
+	//      This means that when full, (kHistorySize - 1) entries will be saved.
+	//      When the table is not full, storing always begins from index 0.
+	int idx = (kHistorySize == _historySize) ? ((_historyIndex + 1) % kHistorySize) : 0;
+	int entriesWritten = 0;
+	while (idx != _historyIndex) {
+		if (!_history[idx].empty()) {
+			saveFile->writeString(_history[idx]);
+			saveFile->writeByte('\n');
+		}
+		idx = (idx + 1) % kHistorySize;
+		++entriesWritten;
```

Comment:
Maybe move this `entriesWritten` increment inside the `if`? To be the actual number of entries written?
It doesn't really matter since this is only used for the debug print. And this will probably not change anything since as discussed in the comments above the empty check is probably not needed. But this seems cleaner to me.

Reply:
This should be fixed in the newer commit https://github.com/scummvm/scummvm/pull/3804/commits/f58aa17c4eead86f0116be320676aff5f5fc9c90